spot-contracts/contracts/vaults/RolloverVault.sol

```diff
@@ -0,0 +1,517 @@
+    // Construction & Initialization
+
+    /// @notice Contract state initialization.
+    /// @param name ERC-20 Name of the vault token.
+    /// @param symbol ERC-20 Symbol of the vault token.
+    /// @param underlying_ ERC-20 address of the underlying token.
+    /// @param perp_ ERC-20 address of the perpetual tranche rolled over.
+    function init(
+        string memory name,
+        string memory symbol,
+        IERC20Upgradeable underlying_,
+        IPerpetualTranche perp_
+    ) public initializer {
+        __ERC20_init(name, symbol);
+        __Ownable_init();
+
+        underlying = underlying_;
+        _syncAsset(underlying_);
+
+        if (perp_.collateral() != underlying_) {
+            revert UnexpectedAsset(underlying_);
+        }
+        perp = perp_;
+    }
+
+    //--------------------------------------------------------------------------
+    // ADMIN only methods
+
+    /// @notice Pauses deposits, withdrawals and vault operations.
+    /// @dev NOTE: ERC-20 functions, like transfers will always remain operational.
+    function pause() public onlyOwner {
+        _pause();
+    }
+
+    /// @notice Unpauses deposits, withdrawals and vault operations.
+    /// @dev NOTE: ERC-20 functions, like transfers will always remain operational.
+    function unpause() public onlyOwner {
+        _unpause();
+    }
+
+    /// @param token The token address.
+    /// @param to The destination address.
+    /// @param amount The amount of tokens to be transferred.
+    function transferERC20(
+        IERC20Upgradeable token,
+        address to,
+        uint256 amount
+    ) external onlyOwner {
+        if (isVaultAsset(token)) {
+            revert UnauthorizedTransferOut(token);
+        }
+        token.safeTransfer(to, amount);
+    }
+
+    //--------------------------------------------------------------------------
+    // External & Public write methods
+
+    /// @notice Recovers deployed funds and redeploys them.
+    /// @dev Simply batches the `recover` and `deploy` functions.
+    function recoverAndRedeploy() external whenNotPaused {
+        recover();
+        deploy();
+    }
+
+    /// @notice Deploys deposited funds.
+    function deploy() public nonReentrant whenNotPaused {
+        IBondController depositBond = perp.getDepositBond();
+        _tranche(depositBond);
+        if (_rollover(perp, depositBond) == 0) {
+            revert NoDeployment();
+        }
+    }
+
+    /// @notice Recovers deployed funds.
+    function recover() public nonReentrant whenNotPaused {
+        _redeemTranches();
+    }
+
+    /// @notice Deposits the provided asset from {msg.sender} into the vault and mints notes.
+    /// @param token The address of the asset to deposit.
+    /// @param amount The amount tokens to be deposited into the vault.
+    /// @return The amount of notes.
+    function deposit(IERC20Upgradeable token, uint256 amount) external nonReentrant whenNotPaused returns (uint256) {
```

Comment:
consider removing the `token` param entirely

Reply:
I wanted to keep the interface generic so this could be a standard for tokenzied vaults. The rollover vault just accepts AMPL, however other vaults could accept different tokens.

Reply:
Sure will do this in a later PR. Adding a todo